pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep.go

```diff
@@ -0,0 +1,118 @@
+)
+
+var (
+	natsOperatorCRDsToDelete = []string{"natsclusters.nats.io", "natsserviceroles.nats.io"}
+)
+
+type kubeClientProvider func(context *service.ActionContext, logger *zap.SugaredLogger) (kubeclient.Client, error)
+
+type removeNatsOperatorStep struct {
+	kubeClientProvider
+}
+
+func newRemoveNatsOperatorStep() *removeNatsOperatorStep {
+	return &removeNatsOperatorStep{
+		kubeClientProvider: defaultKubeClientProvider,
+	}
+}
+
+func defaultKubeClientProvider(context *service.ActionContext, logger *zap.SugaredLogger) (kubeclient.Client, error) {
+	kubeClient, err := kubeclient.NewKubeClient(context.Task.Kubeconfig, logger)
+	if err != nil {
+		return nil, err
+	}
+	return kubeClient, nil
+}
+
+func (r *removeNatsOperatorStep) Execute(context *service.ActionContext, logger *zap.SugaredLogger) error {
+	kubeClient, err := r.kubeClientProvider(context, logger)
+	if err != nil {
+		return err
+	}
+
+	// decorate logger
+	logger = logger.With(log.KeyStep, removeNatsOperatorStepName)
+	// prepare Kubernetes clientset
+	var clientset kubernetes.Interface
+	if clientset, err = context.KubeClient.Clientset(); err != nil {
+		return err
+	}
+
+	// remove the old NATS-operator resources if the NATS-operator deployment still exists
+	natsOperatorDeployment, err := getDeployment(context, clientset, natsOperatorDeploymentName)
+	if err != nil {
+		return err
+	}
+
+	err = r.removeNatsOperatorResources(context, kubeClient, logger, natsOperatorDeployment)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *removeNatsOperatorStep) removeNatsOperatorResources(context *service.ActionContext, kubeClient kubeclient.Client, logger *zap.SugaredLogger, natsOperatorDeployment *appsv1.Deployment) error {
+	if natsOperatorDeployment == nil {
+		logger.With(log.KeyReason, "no nats-operator deployment found").Info("Step skipped")
+		return nil
+	}
+	// get charts from the version when the old NATS-operator yaml resource definitions still exist
+	comp := chart.NewComponentBuilder(natsOperatorLastVersion, natsSubChartPath).
+		WithConfiguration(map[string]interface{}{
+			"global.image.repository": "eu.gcr.io/kyma-project", // replace the missing global value, as we are rendering on the subchart level
+		}).
+		WithNamespace(namespace).
+		Build()
+
+	manifest, err := context.ChartProvider.RenderManifest(comp)
+	if err != nil {
+		return err
+	}
+
+	// set the right eventing name, which went lost after rendering
+	manifest.Manifest = strings.ReplaceAll(manifest.Manifest, natsSubChartPath, eventingNats)
+
+	// remove all the nats-operator resources, installed via charts
+	_, err = context.KubeClient.Delete(context.Context, manifest.Manifest, namespace)
+	if err != nil {
+		return err
+	}
+
+	err = r.removeNatsOperatorCRDs(kubeClient, logger)
+	return err
+}
+
+// delete the leftover CRDs, which were outside of charts
+func (r *removeNatsOperatorStep) removeNatsOperatorCRDs(kubeClient kubeclient.Client, log *zap.SugaredLogger) error {
```

Comment:
I see the CRDs that are removed here on the `cluster-essentials` chart:
https://github.com/kyma-project/kyma/blob/main/resources/cluster-essentials/files/natsserviceroles.nats.crd.yaml
https://github.com/kyma-project/kyma/blob/main/resources/cluster-essentials/files/natsclusters.nats.crd.yaml

Therefore I am not sure if they need to be removed here.

Reply:
These CRDs are outdated and will be removed as a part of this issue as well: https://github.com/kyma-project/kyma/pull/12579